### PR TITLE
faster ftrl fit and predict based on many c-style optimization

### PR DESCRIPTION
**Pull request by stegben**
I worked on some cython optimization on `fit` and `predict` function. The speed is 2X~5X when `interaction=False`:

### before
```
       3 function calls in 4.597 seconds

   Ordered by: standard name

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.000    0.000    0.000    0.000 base.py:99(get_shape)
        1    0.000    0.000    0.000    0.000 {method 'disable' of '_lsprof.Profiler' objects}
        1    4.597    4.597    4.597    4.597 {method 'fit' of 'kaggler.online_model.ftrl.FTRL' objects}

```

### after
```
         3 function calls in 0.981 seconds

   Ordered by: standard name

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.000    0.000    0.000    0.000 base.py:99(get_shape)
        1    0.000    0.000    0.000    0.000 {method 'disable' of '_lsprof.Profiler' objects}
        1    0.981    0.981    0.981    0.981 {method 'fit' of 'kaggler.online_model.ftrl.FTRL' objects}
```

Note that when `interaction=True`, the main overhead will be:
```python
abs(hash('{}_{}'.format(x[i], x[j])))
```
I've provide the solution for it and commented it out in the `_indices` function. I've try that out and find it really solve the overhead. Plz take a look and see if it works general enough.

### CODE REVIEW

**Comment by stegben** [comment]
by the way, here's my profiling code:
```python
import pickle as pkl
import cProfile

import numpy as np
np.random.seed(1234)
import scipy.sparse as sps

from kaggler.online_model import FTRL


DATA_NUM = 4e6


def main():
    print('create y...')
    y = np.random.randint(0, 1, DATA_NUM)
    print('create x...')
    row = np.random.randint(0, 1000000, DATA_NUM)
    col = np.random.randint(0, 100, DATA_NUM)
    data = np.ones(DATA_NUM)
    x = sps.csr_matrix((data, (row, col)), dtype=np.int8)

    print('train...')
    profiler = cProfile.Profile(subcalls=True, builtins=True, timeunit=0.001,)
    clf = FTRL(interaction=False)
    profiler.enable()
    clf.fit(x, y)
    profiler.disable()
    profiler.print_stats()
    clf.predict(x)


if __name__ == '__main__':
    main()
```

File: kaggler/online_model/ftrl.pyx [hunk 1]
```diff
@@ -81,24 +82,27 @@ cdef class FTRL:
...
-            l = len(x)
-            x = sorted(x)
-            for i in xrange(l):
-                for j in xrange(i + 1, l):
-                    yield abs(hash('{}_{}'.format(x[i], x[j]))) % self.n
+            for i in range(x_len):
+                for j in range(i + 1, x_len):
+                    indices.append(abs(hash('{}_{}'.format(x[i], x[j]))) % self.n)
+                    # a much faster and also reasonable way:
+                    # indices.append((x[i] * x[j]) % self.n)
```
Comment (jeongyoonlee) [inline kaggler/online_model/ftrl.pyx]: With this, there are collisions for low indices. e.g. when `x[i] == 1` and `x[j] == 2`, `(x[i] * x[j]) % self.n == 2`.  Alternatively, we may look for a faster `hash` function.

Comment (stegben) [inline kaggler/online_model/ftrl.pyx]: Well, you're right. Randomness for interaction features is needed. Let's deal with this issue in other PR.

Comment (stegben) [inline kaggler/online_model/ftrl.pyx]: Maybe try some simple hash function like Knuth's multiplicative method?

Comment (jeongyoonlee) [inline kaggler/online_model/ftrl.pyx]: Looks promising. As you said, let's deal with it in a separate PR.  Thanks!

Comment (stegben) [inline kaggler/online_model/ftrl.pyx]: Sure, I'll update them after work!

Comment (stegben) [inline kaggler/online_model/ftrl.pyx]: Done. #23

**Pull request merged**